### DISCUSSION TITLE
`goto` in `or` block

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#18618] [Ideas]
To be able to use `goto` statement in the "or" block.

Currently, `goto` can be used in `or` block in this way, but it is a bit semantically odd.
```v
fn foo() ?int{
	return none
}

fn main(){
	i := foo() or {
		unsafe{ goto skip }
		0
	}
	dump(i)
	skip:
	dump('skipped')
}
```
Semantically, it should be this way.
```v
fn main(){
	i := foo() or {
		unsafe{ goto skip }
	}
	dump(i)
	skip:
	dump('skipped')
}
```

Some people say that the `or` block must return a value, but in a `for` loop you can use `break` and `continue` in the `or` block, but they do not return a value.

It is the same concept.

**Comment by spytheman** [comment]
The `or {}` should either return a value, or ensure that execution will not continue further, where the value of `i` has the potential to be used.

In for loops, `continue` and `break` do ensure that `i` will not be used, `continue` by entering a new iteration of the loop body, and `break` by exiting the scope of the loop, so `i` can not be accessed in the outer one.

With goto, you do not have that guarantee, since you can goto a label in the same scope, while the code after the label could be using `i`, and proving that it does not, is not cheap.

**Comment by spytheman** [comment]
That said, given that it is in `unsafe{}`, and that it is the last statement, it does change the control flow, so it can be allowed without the `0`, with perhaps just a notice.

**Comment by ghost** [comment]
I see, indeed, the value cannot be guaranteed.
This is also the user's responsibility.

But having said that, this is not a syntax problem either.
It is, of course, a runtime error.
```v
fn main(){
	unsafe{ goto label }
	a := 'a'
	label:
	dump(a)
}
```

**Comment by spytheman** [comment]
Yes, exactly.
